### PR TITLE
Fix async message updates and add unit test

### PR DESCRIPTION
**Pull request by sydhds**

### CODE REVIEW

**Review by Leo-Besancon — APPROVED** [review]
Great work! The various tested scenarios are nice

File: massa-execution-worker/src/active_history.rs [hunk 1]
```diff
@@ -83,7 +83,9 @@ impl ActiveHistory {
                     return HistorySearchResult::Present(SetUpdateOrDelete::Set(msg));
                 }
                 Some(SetUpdateOrDelete::Update(msg_update)) => {
-                    current_updates.apply(msg_update.clone());
+                    let mut combined_message_update = msg_update.clone();
```
Comment (damip) [inline massa-execution-worker/src/active_history.rs:86]: I think current_updates does not need to be mut anymore

File: massa-execution-worker/src/active_history.rs [hunk 2]
```diff
@@ -83,7 +83,9 @@ impl ActiveHistory {
...
+                    combined_message_update.apply(current_updates);
+                    current_updates = combined_message_update;
                 }
                 Some(SetUpdateOrDelete::Delete) => return HistorySearchResult::Absent,
```
Comment (damip) [inline massa-execution-worker/src/active_history.rs:90]: how about this ? if the `current_updates` actually creates an element while it was deleted before... can it happen?

File: massa-execution-worker/src/speculative_async_pool.rs [hunk 3]
```diff
@@ -287,7 +287,7 @@ impl SpeculativeAsyncPool {
                 }
                 Present(SetUpdateOrDelete::Update(msg_update)) => {
                     current_changes.entry(message_id).and_modify(|e| {
-                        e.apply(msg_update.clone());
+                        *e = msg_update.clone();
```
Comment (damip) [inline massa-execution-worker/src/speculative_async_pool.rs:290]: why was this changed ?

Comment (Leo-Besancon) [reply massa-execution-worker/src/speculative_async_pool.rs:290]: It did work before, but it technically applied the current_changes twice (the returned msg_update already took them into account, so we directly replace it)

File: massa-execution-worker/src/active_history.rs (hunk 1 above)
Comment (Leo-Besancon) [reply massa-execution-worker/src/active_history.rs:86]: I think it still does because of line 88 :)

**Comment by Leo-Besancon** [comment]
@sydhds unless you think otherwise, I think we can merge this one :)

File: massa-execution-worker/src/speculative_async_pool.rs [hunk 4]
```diff
@@ -287,7 +287,7 @@ impl SpeculativeAsyncPool {
...
                     });
                     return true;
                 }
```
Comment (damip) [inline massa-execution-worker/src/speculative_async_pool.rs:293]: question to make sure: do we properly handle the case where there is only an update in the speculative part, and the message is in the final state ?

Comment (Leo-Besancon) [reply massa-execution-worker/src/speculative_async_pool.rs:293]: Seems like it to me yes, in that case:
- On line 264 we find an Update in the pool
- Then, we find nothing in the active history so we do nothing (at line 294)
- Then, we find the message in the final state and we apply the update.
@sydhds do you agree?
